### PR TITLE
Convert metabase-enterprise/collections plugin to TypeScript

### PR DESCRIPTION
**Pull request by kulyk**
Converts a module with paid collections extensions to TypeScript
This is required to properly migrate the new collection form to `formik` in the scope of #26178

### CODE REVIEW

File: enterprise/frontend/src/metabase-enterprise/collections/constants.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  name: t`Official`,
+  icon: "badge",
+  color: "saturated-yellow",
+  tooltips: {
+    default: t`Official collection`,
+    belonging: t`Belongs to an Official collection`,
+  },
+};
+
+export const AUTHORITY_LEVELS: Record<any, AuthorityLevelConfig> = {
```
Comment (ranquild) [inline enterprise/frontend/src/metabase-enterprise/collections/constants.ts:29]: I believe `Record<string, ...` should work?

File: enterprise/frontend/src/metabase-enterprise/collections/components/CollectionAuthorityLevelIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+export function CollectionAuthorityLevelIcon({
+  collection,
+  tooltip = "default",
+  ...iconProps
+}: Props) {
+  if (isRegularCollection(collection)) {
+    return null;
+  }
+  const level = AUTHORITY_LEVELS[collection.authority_level as any];
```
Comment (ranquild) [inline enterprise/frontend/src/metabase-enterprise/collections/components/CollectionAuthorityLevelIcon.tsx:27]: Can we avoid that cast? Will typing `AUTHORITY_LEVELS` as `Record<string, ...` help?

File: frontend/src/metabase/plugins/index.ts [hunk 3]
```diff
@@ -106,20 +108,37 @@ const AUTHORITY_LEVEL_REGULAR: CollectionAuthorityLevelConfig = {
...
     _collection: Collection,
     _onUpdate: (collection: Collection, values: Partial<Collection>) => void,
-  ) => [],
+  ): AuthorityLevelMenuItem[] => [],
 };
 
-export const PLUGIN_COLLECTION_COMPONENTS = {
+export const PLUGIN_COLLECTION_COMPONENTS: Record<
+  string,
+  React.ComponentType<any>
```
Comment (ranquild) [inline frontend/src/metabase/plugins/index.ts:140]: Can we avoid any here too? Maybe we could omit it or specify the props correctly

**Review by ranquild — APPROVED** [review]
I believe we could avoid using `any` in some places, but otherwise 👍

File: enterprise/frontend/src/metabase-enterprise/collections/components/CollectionAuthorityLevelIcon.tsx (hunk 2 above)
Comment (kulyk) [reply enterprise/frontend/src/metabase-enterprise/collections/components/CollectionAuthorityLevelIcon.tsx:27]: @ranquild yeah, I also don't like that; the thing is that `authority_level` is `null | "official"` and TypeScript doesn't like `null` being a key of a record :/ Do you happen to know a better workaround for this?

Comment (ranquild) [reply enterprise/frontend/src/metabase-enterprise/collections/components/CollectionAuthorityLevelIcon.tsx:27]: Yeah `null` cannot be the key of the object. In runtime it gets converted to string. So we need to do:
```
const AUTHORITY_LEVELS: Record<string, string> = {
  [String(regularCollection.type)]: ...
};
```

Comment (ranquild) [reply enterprise/frontend/src/metabase-enterprise/collections/components/CollectionAuthorityLevelIcon.tsx:27]: And when you access it, `const level = AUTHORITY_LEVELS[String(collection.authority_level)]`

Comment (kulyk) [reply enterprise/frontend/src/metabase-enterprise/collections/components/CollectionAuthorityLevelIcon.tsx:27]: Thanks! Let me try that 👍

**Pull request merged**